Add node.js binding for activations

mingmingtasd opened the pull request.

mingmingtasd commented:
I have verified that related webnn-polyfill tests can pass through node.js binding. PTAL, thanks! @fujunwei

huningxin commented on node/src/GraphBuilder.cpp:
```diff
@@ -61,6 +62,24 @@ Napi::FunctionReference node::GraphBuilder::constructor;
     operand->SetImpl(mImpl.op(input));                                                      \
     return object;
 
+#define BUILD_ACTIVATION_UNARY(op)                                                              \
```
Change this macro to `BUILD_UNARY_OPERATOR` and only keep the code to build operator.

huningxin commented on node/src/GraphBuilder.cpp:
```diff
@@ -61,6 +62,24 @@ Napi::FunctionReference node::GraphBuilder::constructor;
+#define BUILD_ACTIVATION_UNARY(op)                                                              \
+    std::vector<napi_value> args;                                                               \
+    ml::Operand input;                                                                          \
+    if (!IsOperator(info)) {                                                                    \
```
The following code is duplicated with `BUILD_UNARY`, is it?

huningxin commented on node/src/GraphBuilder.cpp:
```diff
@@ -151,15 +170,15 @@ namespace node {
 
     Napi::Value GraphBuilder::Relu(const Napi::CallbackInfo& info) {
-        BUILD_UNARY(Relu);
+        BUILD_ACTIVATION_UNARY(Relu);
```
With `BUILD_UNARY` and `BUILD_UNARY_OPERATOR`, it could be
```c++
if (info.Length() == 0) {
  BUILD_UNARY_OPERATOR(Relu);
} else {
  BUILD_UNARY(Relu);
}
```

huningxin commented on node/src/Utils.h:
```diff
@@ -251,6 +252,22 @@ namespace node {
         return size;
     }
 
+    inline bool IsOperator(const Napi::CallbackInfo& info) {
```
I don't think this is generic enough, for example unary op only needs to check whether `info.Length() == 0`. How about check them case by case?

huningxin commented on node/src/ops/Clamp.cpp:
```diff
@@ -15,27 +15,35 @@
         std::vector<napi_value> args;
         ml::Operand input;
-        WEBNN_NODE_ASSERT(GetOperand(info[0], input, args), "The input parameter is invalid.");
+        if (!IsOperator(info)) {
```
Please check the arguments for Clamp specifically.

huningxin reviewed (commented):
Thanks @mingmingtasd . Please also roll the webnn-polyfill in DEPS that includes the js tests for fused activation.

mingmingtasd commented:
Much thanks for @fujunwei 's help, we have fixed the clamp related issue. I updated the commit id in DEPS for webnn-samples and webnn-polyfill to the latest. PTAL, thanks! @huningxin @fujunwei

huningxin commented on node/src/Operand.cpp:
```diff
@@ -22,10 +22,9 @@ namespace node {
-                                         "The input must be Operand object.");
-            mInputs.push_back(Napi::Persistent(operand));
+            // It may be an operand or an operator.
+            Napi::Object object = info[i].As<Napi::Object>();
```
probably you can still check whether the object is `Operand` or `Operator`, can you?

mingmingtasd replied:
Yes, done,thanks! @huningxin

huningxin reviewed (approved):
LGTM with one nit. Please take a look.

mingmingtasd commented on node/src/ops/Clamp.cpp:
```diff
@@ -15,27 +15,38 @@
         std::vector<napi_value> args;
         ml::Operand input;
-        WEBNN_NODE_ASSERT(GetOperand(info[0], input, args), "The input parameter is invalid.");
+        bool isOperator = info.Length() == 0 ||
```
isOperator is used to check whether the node itself is an operator, right?

fujunwei replied:
The `clamp `in `Operand clamp(Operand x, optional ClampOptions options = {});` is also operator now, so we can distinguish with name `isFusedOperator`

mingmingtasd replied:
Thanks, done, PTAL. @fujunwei

fujunwei reviewed (approved):
LGTM, let's merge.
